### PR TITLE
Alter the parser and introduce new rule

### LINKED ISSUE #1044: zndsl: support subtracting metrics
### Is your feature request related to a problem? Please describe.

Add support for writing an asserting on `a - b` where `a` and `b` are prometheus metrics.

### Describe the solution you'd like

Example:
```
alice: substrate_block_height{status="best"} - substrate_block_height{status="finalized"} is greater than 10
```

### PR DESCRIPTION
**Pull request by wirednkod**
Fixes #1044

### CODE REVIEW

File: crates/parser/src/ast.rs [hunk 1]
```diff
@@ -43,6 +49,16 @@ pub enum AssertionKind {
...
+    CalcMetrics {
+        node_name: NodeName,
+        metric_name: String,
+        math_ops: MathOps,
+        metric_name_calc: String,
+        op: Operator,
+        target_value: u64,
+        #[serde(with = "optional_timeout")]
+        timeout: Option<Duration>,
+    },
```
Comment (pepoviola) [inline crates/parser/src/ast.rs:61]: I think we can call the metrics a/b, something like this

```suggestion
    CalcMetrics {
        node_name: NodeName,
        metric_name_a: String,
        math_ops: MathOps,
        metric_name_b: String,
        op: Operator,
        target_value: u64,
        #[serde(with = "optional_timeout")]
        timeout: Option<Duration>,
    },
```
I think is more clear that `_calc` wdyt?

File: crates/parser/src/lib.rs [hunk 2]
```diff
@@ -333,6 +345,36 @@ pub fn parse(unparsed_file: &str) -> Result<ast::TestDefinition, errors::ParserE
 
                 assertions.push(assertion);
             }
+            Rule::calc_metrics => {
+                // Pairs should be in order:
+                // name, para_id, block_height, minus, finalized_height, [timeout]
```
Comment (pepoviola) [inline crates/parser/src/lib.rs:350]:
```suggestion
                // name, para_id, block_height, sign, finalized_height, [timeout]
```

File: crates/parser/src/lib.rs [hunk 3]
```diff
@@ -333,6 +345,36 @@ pub fn parse(unparsed_file: &str) -> Result<ast::TestDefinition, errors::ParserE
...
+                    Some(parse_within(within_rule)?)
+                } else {
+                    None
+                };
+
+                let assertion = Assertion {
+                    parsed: AssertionKind::CalcMetrics {
+                        node_name: name.to_owned(),
+                        metric_name,
+                        math_ops: math_ops,
```
Comment (pepoviola) [inline crates/parser/src/lib.rs:367]:
```suggestion
                        math_ops,
```
clippy suggestion.

File: crates/parser/src/zombienet.pest [hunk 4]
```diff
@@ -41,7 +41,12 @@ op_lt        = { "<" | "is lower than" }
 op_gt        = { ">" | "is greater than" | "greater than" }
 op_eq        = { "==" | "=" | "equals" | "is equal to" | "is" }
 op_ineq      = { "!=" }
-comparison = { (op_lte | op_gte | op_gt | op_lt | op_eq | op_ineq) ~ int+ }
+comparison   = {(op_lte | op_gte | op_gt | op_lt | op_eq | op_ineq) ~ int+ }
```
Comment (pepoviola) [inline crates/parser/src/zombienet.pest:44]:
```suggestion
comparison = { (op_lte | op_gte | op_gt | op_lt | op_eq | op_ineq) ~ int+ }
```
fmt

File: examples/0001-small-network.zndsl [hunk 5]
```diff
@@ -22,4 +22,8 @@ alice: system event matches glob "*was backed*" within 10 seconds
 
 #parachain tests
 alice: parachain 100 is registered within 225 seconds
-alice: parachain 100 block height is at least 10 within 200 seconds
\ No newline at end of file
+alice: parachain 100 block height is at least 10 within 200 seconds
+
+## test the block height - (or minus) finalised block
+alice: reports block height minus finalised block is lower than 10
+alice: reports block height - finalised block is lower than 10
```
Comment (pepoviola) [inline examples/0001-small-network.zndsl:29]:
```suggestion
alice: reports block height - finalised block is lower than 10 within 200 seconds
```
We should test version with `within` keyword.

File: javascript/packages/orchestrator/src/test-runner/assertions.ts [hunk 6]
```diff
@@ -79,6 +79,50 @@ const Report = ({
...
+    const promise2 = nodes.map((node: any) =>
+      node.getMetric(
+        metric_name_calc!,
+        toChaiComparator(op!),
+        target_value!,
+        timeout || DEFAULT_INDIVIDUAL_TEST_TIMEOUT,
+      ),
+    );
+
+    const values = await Promise.all([...promise1, ...promise2]);
```
Comment (pepoviola) [inline javascript/packages/orchestrator/src/test-runner/assertions.ts:112]: In theory we can have `groups` here where we get more than two values, that why we get an array of promises.
We should cover that case by looping (by two) the `values` and checking that for every pair the comparison success.

**Review by pepoviola — COMMENTED** [review]
Looks good! A couple of comments to discuss.
Thanks!!

**Review by pepoviola — APPROVED** [review]
Awesome, thanks!!

File: javascript/packages/orchestrator/src/test-runner/assertions.ts [hunk 7]
```diff
@@ -89,6 +89,15 @@ const CalcMetrics = ({
...
 }: FnArgs) => {
   const comparatorFn = comparators[op!];
+  const mathFn =
+    math_ops === "Minus"
+      ? (a: number, b: number): number => {
+          return a - b;
+        }
+      : (a: number, b: number): number => {
+          return a + b;
+        };
```
Comment (l0r1s) [inline javascript/packages/orchestrator/src/test-runner/assertions.ts:99]: You could use the math_ops value inside the closure ?

```suggestion
        (a: number, b: number): number => {
          return math_ops === "Minus ? a - b : a + b;
        }
```

**Review by l0r1s — APPROVED** [review]

**Comment by l0r1s** [comment]
:+1:

**Pull request merged**